catwalk/models/gpt.py

```diff
@@ -1,34 +1,164 @@
+            AutoModelForCausalLM, self.pretrained_model_name_or_path, False).eval().to(device)
+        tokenizer = cached_transformers.get_tokenizer(
+            AutoTokenizer, self.pretrained_model_name_or_path)
+        tokenizer.pad_token = tokenizer.eos_token
+        tokenizer.padding_side = 'left'
+        
+        converted_instances = self._convert_instances(instances, InstanceFormat.HF_QA, task)
+
+        def format_instance(instance: HFQAInstance) -> Tuple[str, str]:
+            # TODO: Use promptsource to add more prompt options?
+            return instance.context, f"\nQuestion:{instance.question}\nAnswer:"
+
+        converted_instances = Tqdm.tqdm(converted_instances, desc="Processing instances")
+        for batch in more_itertools.chunked(converted_instances, batch_size):
+            formatted_batch = [format_instance(instance) for instance in batch]
+            encodings = tokenizer(formatted_batch,
+                                  padding="longest",
+                                  truncation="only_first",
+                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],  # TODO: Make this configurable?
+                                  stride=128,
+                                  return_overflowing_tokens=True,
+                                  return_offsets_mapping=True,
+                                  return_token_type_ids=True,
+                                  return_tensors="pt"
+                                  )
+            sample_map = encodings.pop("overflow_to_sample_mapping")
+            offset_mapping = encodings.pop("offset_mapping")
+            token_type_map = encodings.pop("token_type_ids")
+
+            # each example in the batch may have multiple sampled windows due to truncation
+            example_idx_to_sample_idx = defaultdict(list)
+            for i, example_idx in enumerate(sample_map):
+                example_idx_to_sample_idx[example_idx.item()].append(i)
+
+            filtered_encodings: Dict[str, List] = {"input_ids": [], "attention_mask": []}
+
+            # We need to use the offset mapping to get the actual start and end indices of the context.
+            # To do this we need to find the index of the first/last token in the context.
+            # For example,
+            # Tokens:         PADDING CONTEXT QUESTION
+            # Token Type IDs:    0        0       1 (Used to find the last token of the context)
+            # Attention Mask:    0        1       1 (Used to find the first token of the context)
+
+            def get_token_bounds_for_context(sample_idx: int) -> Tuple[int, int]:
+                attention_mask = encodings["attention_mask"][sample_idx]
+                sequence_ids = token_type_map[sample_idx]
+
+                start_index = 0
+                while (attention_mask[start_index] == 0):
+                    start_index += 1
```

Comment:
Yes I found that doing this improves performance especially if we are not fine-tuning the model like in in-context learning.

Reply:
Improves quality of the answers, or speed?

Reply:
Quality of answer

Reply:
I think standard approach is to pad on the right, but adjust position ids? @jagnusson, didn't you do this for OPT?

We want to pretend that each sequence starts at position 0, and that context and continuation are directly adjacent. When the contexts in a batch aren't all the same size, the only way we can do both things is by messing with position ids.

Reply:
> I think standard approach is to pad on the right, but adjust position ids? @jagnusson, didn't you do this for OPT?
> 
> We want to pretend that each sequence starts at position 0, and that context and continuation are directly adjacent. When the contexts in a batch aren't all the same size, the only way we can do both things is by messing with position ids.

Sorry I didn't see this before! I need to tinker with my notification settings more apparently.

So I unfortunately I don't actually have first hand experience with this. The Metaseq OPT provides its own padding strategy and you just pass it a list of unpadded inputs.

When I looked into this it seemed like [this discussion](https://discuss.huggingface.co/t/batch-generation-with-gpt2/1517) was the definitive one on batched generation in HF. They use padding from the left. I have the same question as @dirkgr though cuz I don't know how they handle the position id's then. I'm not sure if there is an elegant way to pad from the right though because the attention mask may not play nice with trying to mask out padding tokens that fall between the context and the generated continuation (and you still need to adjust the position ids if you want to do that).